src/rpc/mod.rs

```diff
@@ -2,46 +2,133 @@
 // SPDX-License-Identifier: Apache-2.0, MIT
 
 mod auth_api;
+mod auth_layer;
 mod beacon_api;
 mod chain_api;
 mod common_api;
+mod error;
 mod eth_api;
 mod gas_api;
 mod mpool_api;
 mod net_api;
 mod node_api;
-mod rpc_http_handler;
-mod rpc_util;
-mod rpc_ws_handler;
 mod state_api;
 mod sync_api;
 mod wallet_api;
 
+use std::error::Error as StdError;
+use std::net::SocketAddr;
 use std::sync::Arc;
 
+use crate::key_management::KeyStore;
+use crate::rpc::auth_layer::AuthLayer;
+use crate::rpc::{
+    beacon_api::beacon_get_entry,
+    common_api::{session, shutdown, start_time, version},
+    state_api::*,
+};
 use crate::rpc_api::{
     auth_api::*, beacon_api::*, chain_api::*, common_api::*, data_types::RPCState, eth_api::*,
     gas_api::*, mpool_api::*, net_api::*, node_api::NODE_STATUS, state_api::*, sync_api::*,
     wallet_api::*,
 };
-use axum::routing::{get, post};
+
 use fvm_ipld_blockstore::Blockstore;
-use jsonrpc_v2::{Data, Error as JSONRPCError, Server};
-use tokio::net::TcpListener;
+use hyper::server::conn::AddrStream;
+use hyper::service::{make_service_fn, service_fn};
+use jsonrpc_v2::Error as JSONRPCError;
+use jsonrpsee::server::{
+    stop_channel, RpcModule, RpcServiceBuilder, Server, StopHandle, TowerServiceBuilder,
+};
+use jsonrpsee::Methods;
 use tokio::sync::mpsc::Sender;
+use tokio::sync::RwLock;
+use tower::Service;
 use tracing::info;
 
-use crate::rpc::{
-    beacon_api::beacon_get_entry,
-    common_api::{session, shutdown, start_time, version},
-    rpc_http_handler::{rpc_http_handler, rpc_v0_http_handler},
-    rpc_ws_handler::{rpc_v0_ws_handler, rpc_ws_handler},
-    state_api::*,
-};
+const MAX_RESPONSE_BODY_SIZE: u32 = 16 * 1024 * 1024;
+
+#[derive(Clone)]
+struct PerConnection<RpcMiddleware, HttpMiddleware> {
+    methods: Methods,
+    stop_handle: StopHandle,
+    svc_builder: TowerServiceBuilder<RpcMiddleware, HttpMiddleware>,
+    keystore: Arc<RwLock<KeyStore>>,
+}
 
 pub async fn start_rpc<DB>(
-    state: Arc<RPCState<DB>>,
-    rpc_endpoint: TcpListener,
+    state: RPCState<DB>,
+    rpc_endpoint: SocketAddr,
+    forest_version: &'static str,
+    shutdown_send: Sender<()>,
+) -> Result<(), JSONRPCError>
+where
+    DB: Blockstore + Send + Sync + 'static,
+{
+    // `Arc` is needed because we will share the state between two modules
+    let state = Arc::new(state);
+    let keystore = state.keystore.clone();
+    let mut module = RpcModule::new(state.clone());
+
+    register_methods(
+        &mut module,
+        state.state_manager.chain_config().block_delay_secs as u64,
```

Comment:
nit: `u64::from(...)` so readers don't need to ask themselves if this will truncate

Reply:
Yes, that's way better.